doc/developer/design/20230214_transform_using.md

```diff
@@ -0,0 +1,209 @@
+miss the original update from `RAW` and we will fail to produce the differential update that we
+would have produced had we not crashed.
+
+In order to solve this we will need to decompose `Q_mfp` into three parts: `Q_safemfp`,
+`Q_temporal_filter`, `Q_final_projection`
+
+> 💡 To understand the following decomposition it is important to keep in mind that all temporal
+> predicates are internally converted into expressions that produce a candidate timestamp for either
+> the lower or upper bound of the validity period. For example, the predicate `expr <= mz_now()`
+> will get converted into just the expression `expr` and will be stored in a `lower_bounds` list. To
+> calculate the lower and upper bounds of the validity period of a row during a temporal MFP
+> evaluation all the lower bound expressions are evaluated and the maximum is selected and all the
+> upper bound expressions are evaluated and minimum is selected.
+
+The goal of the decomposition is to come up with a `Q_safemfp` that contains no temporal filters,
+performs the maximum amount of mapping and filtering, and retains enough information to later
+recover the original validity period. We can define the three parts like so:
+
+1. `Q_safemfp` will contain the non-temporal part of `Q_mfp` combined with two additional `Map`
+   expressions. These two additional map expressions will correspond to the lower and upper bounds
+   and will be calculating the final timestamp for them. We can do so by fabricating two
+   `MirScalarExpr`s. One that calculates the maximum of all the lower bound exprs and one that
+   calculates the minimum of all the upper bound exprs. Effectively, we will precompute the validity
+   period and store it as two additional timestamp columns at the end of the original row.
+2. `Q_temporal_filter` will contain two trivial temporal predicates:
+    1. `lower_ts <= mz_now()`
+    2. `mz_now() <= upper_ts`
+3. `Q_final_projection` will be simply throwing away the two last columns (`lower_ts` and
+   `upper_ts`) of the row
+
+Given this decomposition we can now define the ingestion as:
+
+1. `C_intermediate = PERSIST(Q_temporal_filter(Q_safemfp(RAW)))`
+2. `C = PERSIST(Q_expr(Q_final_projection(Q_temporal_filter(C_intermediate))))`
+
+### Planner/Optimizer/Data structure stabilization
+
+The inability to have `RAW` snapshots readily available has some deep implications on how much of
+our internal choices we are willing to stabilize, which in turn have deep implications on the kind
+of queries we want to allow users to use in the `TRANSFORM USING` position.
+
+In the equations above we had to define `C_intermediate` to effectively be a collection whose schema
+is very much dependent on planner and optimizer choices. Since the transformation from `RAW` to
+`C_intermediate` is not described by high level `SQL` that can be stored in the catalog we must
+explore what else we must stabilize in order to be able to continue ingesting a source across
+versions.
```

Comment:
Maybe another option is to restrict our attention to transforms that can be represented in SQL. For example, extracting `WHERE` clauses that contain only temporal upper bounds, or entire fragments (projections and all) that contain no temporal lower bounds. I like this less, personally, but if we were allergic to stabilizing anything (or didn't want to wait) we could potentially start with this (pushing down some subset of filters, and transforms/projections only if lower bounds do not exist).